apps/builder/app/builder/features/settings-panel/variables.tsx

```diff
@@ -108,38 +115,46 @@ const saveVariable = (
     return;
   }
   const [instanceId] = instanceSelector;
-  serverSyncStore.createTransaction([$dataSources], (dataSources) => {
-    let variableValue: Extract<DataSource, { type: "variable" }>["value"] = {
-      type: "json",
-      value,
-    };
-    if (typeof value === "string") {
-      variableValue = { type: "string", value };
-    }
-    if (typeof value === "number") {
-      variableValue = { type: "number", value };
-    }
-    if (typeof value === "boolean") {
-      variableValue = { type: "boolean", value };
+  serverSyncStore.createTransaction(
+    [$dataSources, $resources],
+    (dataSources, resources) => {
+      // cleanup resource when value variable is set
+      if (dataSource?.type === "resource") {
+        resources.delete(dataSource.resourceId);
+      }
+
+      let variableValue: Extract<DataSource, { type: "variable" }>["value"] = {
+        type: "json",
+        value,
+      };
+      if (typeof value === "string") {
+        variableValue = { type: "string", value };
+      }
+      if (typeof value === "number") {
+        variableValue = { type: "number", value };
+      }
+      if (typeof value === "boolean") {
+        variableValue = { type: "boolean", value };
```

Comment:
this feels like it shold be generalized in a utility somewhere, I think I saw similar code already before